### ISSUE TITLE
Complete macro implementation from rfc

### ISSUE THREAD

**Issue by ef4** [embroider-build/embroider#397]
This tracks bringing `@embroider/macros` up to date with [RFC 507](https://github.com/emberjs/rfcs/pull/507).

Here's a table to track progress.
 - "implemented macro" means that the build-time implementation is done and matches the RFC. Some of the unchecked ones may have implementations but need review to compare against the latest RFC
 - "implemented types & runtime" means that the run-time implementation is done and matches the RFC. This also provides the typescript types that users of the macros will see.

| name                             | type| implemented macro | implemented types & runtime | documented |
| --------------------- |----- |---------------------|-------------------------------|--------------|
| dependencySatisfies | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| each                             | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| failBuild                        | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| getConfig                    | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| getOwnConfig              | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| importSync                  | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| macroCondition         | js      | ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| moduleExists              | js      |  ✔️                               |                                                       |                        |
|                                       | hbs   |                                     |                                                       |                        |
| maybeAttrs                 | js      | not applicable             |   not applicable                          |  not applicable |
|                                       | hbs   |                                     |                                                       |                        |

**Comment by simonihmig** [comment]
Here's a first attempt at updating the hbs macros: #404

**Comment by rwjblue** [comment]
The grid probably needs tweaking now that #404 was landed.

**Comment by simonihmig** [comment]
> The grid probably needs tweaking now that #404 was landed.

Yes. As [said before](https://github.com/embroider-build/embroider/pull/404#issuecomment-603428496) I wanted to do that, but was a bit unsure if I missed anything from the RFC. Assuming this is not the case, I updated the grid now!

**Comment by simonihmig** [comment]
I updated the grid further with what my understanding is of the runtime implementations, but please double check!
* Babel:
  * `failBuild`, `dependencySatisfies`, `importSync` don't have a run-time implementation (other than a TS stub that throws)
  * `getConfig`, `getOwnConfig`, `each` and `macroCondition` have a run-time implementation already
  * `moduleExists` is still missing, which should also be a stub that throws, right?
* Glimmer: 
  * no runtime implementations yet at all
  * `macroDependencySatisfies` and `macroFailBuild` will be build-time only, and as such not have any run-time implementation
